multi_asset/multi_asset.md

```diff
@@ -221,12 +221,20 @@ and balance.
 #### Approval
 
 The entry points `Add_operator`/`Remove_operator` allow an operator to manage
-one’s entire set of tokens on behalf of the approver. To permit approval of a
+one’s entire set of tokens on behalf of the approver. To approve management of a
 subset of token IDs, an interface such as
 [ERC-1761 Scoped Approval Interface](https://eips.ethereum.org/EIPS/eip-1761)
 is suggested. The counterpart `Is_operator` provides introspection into
 any status set by ``Add_operator`/`Remove_operator`.
 
+Only token owner contract can invoke `Add_operator`/`Remove_operator` entry
+points on multi asset contract and manage its operators. Token owner contract
+MUST implement `multi_token_receiver` interface. If the owner does not implement
+`multi_token_receiver` interface, `Add_operator` SHOULD fail.
```

Comment:
In particular, I ask yet again, what does "MUST implement `multi_token_receiver` interface" mean?

There is no such concept of "implementing an interface" in Michelson. (I claim there is no such concept in LIGO either!)

What the spec should say is that the receiver must have a %on_multi_tokens_received entrypoint with a type which is compatible with the expected type (which should be given as a Michelson type, but, that is a separate story.) It is up to you to decide what "compatible" should mean exactly, here, and the spec should say that. (You could use Michelson's definition of "compatibility", but it might be a little bit too lax. You could say that the type must exactly match the expected type. There are potential problems with both answers, but, either way is fine by me.

The spec should say that the token contract must enforce this by (or as if) using `CONTRACT %on_multi_tokens_received ...`. A token contract which uses `CONTRACT ...` (like the draft implementation) will violate the spec, and will be dangerously incorrect.

Reply:
There is implicit notion of the interface with Michelson multi-entry points with nested annotations.

If someone define an "interface" `A` as
```
type A =
  | A_entry1 of t1
  | A_entry2 of t2
```

the contract "implements" this interface if it has entry point `(or %a (t1 %a_entry1) (t2 %a_entry2))`

Alternatively, you can use "duck typing" semantics: if a contract has entry points `%a_entry1` and `%a_entry2`, it implements interface `A`